workflow/certprof/context.go

```diff
@@ -0,0 +1,76 @@
+package certprof
+
+import (
+	"encoding/json"
+	"errors"
+)
+
+var (
+	ErrNilContext       = errors.New("nil context")
+	ErrEmptyProfileName = errors.New("empty profile name provided")
+	ErrEmptyFilter      = errors.New("empty filter")
+)
+
+// Filter filters the returned CertificateList certificates by maching attributes.
+type Filter struct {
+	// CNPrefix matches certificates based on the CN prefix matching this string.
+	CNPrefix string `json:"cn_prefix,omitempty"`
+
+	// NoRequireIdentity disables the requirement of the certificate
+	// to be marked as an Identity in the CertificateList response.
+	NoRequireIdentity bool `json:"require_identity,omitempty"`
+}
+
+// Criteria specifies how to check that certificate should be replaced.
+type Criteria struct {
+	// AlwaysReplace is just that: a forcing mechanism to always replace the certificate.
+	AlwaysReplace bool `json:"always_replace,omitempty"`
+
+	// UntilExpirySeconds checks that the certificate NotAfter timestamp is still valid.
+	// I.e. this is intended to check that a certificate should be replaced
+	// before it expires.
+	// E.g. a value of 300 would mean that a certificate should be replaced
+	// if the certificate NotAfter time is within 5 minutes of expiring (or already expired).
+	UntilExpirySeconds int `json:"until_expiry_sec,omitempty"`
+}
+
+// Context configures workflow behavior.
+type Context struct {
+	// NoManagedOnly turns off the default behavior setting "ManagedOnly" to
+	// true in the CertificateList command.
+	NoManagedOnly bool `json:"no_managed_only,omitempty"`
```

Comment:
Instead of `NoManagedOnly` maybe `AllCertificates` might be more clear. It would change it to:

"AllCertificates turns off the default behavior of setting "ManagedOnly" to true in the CertificateList command"

Reply:
It would seem more clear, I agree. However I didn't want to speak something affirmative that Apple didn't intend. I.e:

https://github.com/apple/device-management/blob/2e6673873e71a8bba2516f915710e5bd50b81859/mdm/commands/certificate.list.yaml#L47-L65

Notably this passage:

> User-enrolled devices ignore this value and always only include managed certificates.

So by saying "all certificates" we'd be implying something that isn't true. Plus I like that it "calls back" to Apple's docs as the source of truth.

But yes, I agree, the double-negative, in general, is not as clear as it could be.